### ISSUE TITLE
Support self signed certs for alert notifications

### ISSUE THREAD

**Issue by laughland** [grafana/grafana#9548]
- What Grafana version are you using? `grafana/grafana:master`
- What datasource are you using? `Prometheus`
- What OS are you running grafana on? `CentOS 7 on virtualBox` 
- What did you do? `Tried to add a notification alert channel via the grafana UI`
- What was the expected result? `A notification would be sent`
- What happened instead? 

```
grafana_1        | t=2017-10-16T18:06:24+0000 lvl=eror msg="Failed to send alert notifications" logger=context userId=1 orgId=1 uname=admin error="Post https://mattermost.XXXXXXXXXX.com/hooks/<hook id>: x509: certificate signed by unknown authority"
grafana_1        | t=2017-10-16T18:06:24+0000 lvl=eror msg="Request Completed" logger=context userId=1 orgId=1 uname=admin method=POST path=/api/alert-notifications/test status=500 remote_addr=X.X.X.X time_ms=26 size=48 referer=http://localhost:3000/alerting/notification/1/edit
 ```

To troubleshoot we manually added our certificates with Grafana running in a Docker container:

 `docker run --rm --entrypoint="/bin/bash" -p 3000:3000 -it grafana/grafana`

Our signed certificates were added to `/etc/ssl/certs` and then we ran `c_rehash /etc/ssl/certs`

We curled our mattermost instance `curl https://mattermost.ourdomain.com` successfully:

```
root@4b3ababf117c:/# curl https://mattermost.ourdomain.com
<!DOCTYPE html> <html> <head> <meta http-equiv=X-UA-Compatible content="IE=edge"> <meta name=viewport content="width=device-width,initial-scale=1,maximum-scale=1,user-scalable=0"> ....
```

Is there some way to use our self signed certificates with Grafana to add an alerting channel?

**Comment by bergquist** [comment]
There are currently no way of using self signed certs for alert notifications. 

https://github.com/grafana/grafana/blob/master/pkg/services/notifications/webhook.go#L27

**Comment by bergquist** [comment]
@laughland Btw. Im not sure your using the correct path for certs on Centos. https://unix.stackexchange.com/questions/363308/how-do-i-list-all-available-ssl-ca-certificates-on-centos-6/363309#363309

**Comment by laughland** [comment]
Thanks for the quick reply. I tried your suggestions via the links but I with no luck; we cannot get past the `x509: certificate signed by unknown authority`  error.  I'm including more details. 

In my `docker-compose` file I have this section for Prometheus and Alertmanager:
```yaml
 services:
prometheus:
     image: prom/prometheus:latest
     volumes:
       - ./prometheus/:/etc/prometheus/
       - prometheus_data:/prometheus
       #- ./prometheus_snmp/certs/:/etc/pki/tls/certs/
       - /alertmanager/alert.rules:/etc/alertmanager/alert.rules
     command:
       - '--config.file=/etc/prometheus/prometheus.yml'
       - '--storage.local.path=/data'
       - '--alertmanager.url=http://localhost:9093/'
       - '--web.external-url=http://localhost:9090/'

```
#I have a non-self signed cert in the volume defined in the above #`- ./prometheus_snmp/certs/`'
 

```yaml

 alert-manager:
       image: prom/alertmanager:latest
       volumes:
         - ./alertmanager/:/etc/alertmanager/
       command:
         - '--config.file=/etc/alertmanager/alertmanager.yml'
```



In my `prometheus.yml` file I have:

```yaml
global:
  scrape_interval: 500s
  scrape_timeout: 500s

alerting:
  alertmanagers:
  - tls_config:
      insecure_skip_verify: true
      #   ca_file: '/etc/pki/tls/certs/ca-bundle.crt'
    static_configs:
      - targets:
```

I  tried using `insecure_skip_verify: true` but no matter how I have tried to configure this I still get this output when running docker-compose:

```javascript
alert-manager_1  | time="2017-10-31T14:37:47Z" level=error 
msg="Notify for 1 alerts failed: Post https://mattermost.<domain>.com/hooks/<key>:
 x509: certificate signed by unknown authority" source="dispatch.go:264" 
```

We have poured through the docs and still cannot get this to work. (https://prometheus.io/docs/operating/configuration/#%3Cfilename%3E)


Here are details on Docker, Prometheus, Alertmanager versions/tags:

Docker:
```bash
# docker version
Client:
 Version:      17.05.0-ce
 API version:  1.29
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:06:25 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.05.0-ce
 API version:  1.29 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:06:25 2017
 OS/Arch:      linux/amd64
 Experimental: false

```

Prometheus and Alertmanager are using `prom/prometheus:latest` and  `prom/alertmanager:latest` respectively.

Any further details on how to set this up are appreciated. Thanks again.

**Comment by laughland** [comment]
Thanks for the quick reply. I tried your suggestions via the links but I with no luck; we cannot get past the `x509: certificate signed by unknown authority`  error.  I'm including more details. 

In my `docker-compose` file I have this section for Prometheus and Alertmanager:
```yaml
 services:
prometheus:
     image: prom/prometheus:latest
     volumes:
       - ./prometheus/:/etc/prometheus/
       - prometheus_data:/prometheus
       #- ./prometheus_snmp/certs/:/etc/pki/tls/certs/
       - /alertmanager/alert.rules:/etc/alertmanager/alert.rules
     command:
       - '--config.file=/etc/prometheus/prometheus.yml'
       - '--storage.local.path=/data'
       - '--alertmanager.url=http://localhost:9093/'
       - '--web.external-url=http://localhost:9090/'

```
#I have a non-self signed cert in the volume defined in the above #`- ./prometheus_snmp/certs/`'
 

```yaml

 alert-manager:
       image: prom/alertmanager:latest
       volumes:
         - ./alertmanager/:/etc/alertmanager/
       command:
         - '--config.file=/etc/alertmanager/alertmanager.yml'
```



In my `prometheus.yml` file I have:

```yaml
global:
  scrape_interval: 500s
  scrape_timeout: 500s

alerting:
  alertmanagers:
  - tls_config:
      insecure_skip_verify: true
      #   ca_file: '/etc/pki/tls/certs/ca-bundle.crt'
    static_configs:
      - targets:
```

I  tried using `insecure_skip_verify: true` but no matter how I have tried to configure this I still get this output when running docker-compose:

```javascript
alert-manager_1  | time="2017-10-31T14:37:47Z" level=error 
msg="Notify for 1 alerts failed: Post https://mattermost.<domain>.com/hooks/<key>:
 x509: certificate signed by unknown authority" source="dispatch.go:264" 
```

We have poured through the docs and still cannot get this to work. (https://prometheus.io/docs/operating/configuration/#%3Cfilename%3E)


Here are details on Docker, Prometheus, Alertmanager versions/tags:

Docker:
```bash
# docker version
Client:
 Version:      17.05.0-ce
 API version:  1.29
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:06:25 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.05.0-ce
 API version:  1.29 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:06:25 2017
 OS/Arch:      linux/amd64
 Experimental: false

```

Prometheus and Alertmanager are using `prom/prometheus:latest` and  `prom/alertmanager:latest` respectively.

Any further details on how to set this up are appreciated. Thanks again.

**Comment by bergquist** [comment]
I think you closed this by mistake. re opening.

**Comment by laughland** [comment]
Yes, thanks!

**Comment by andrewlarioza** [comment]
I'm having the same issue with our Grafana Alerts channel using webhook. The webhook url is also using a self-signed ssl certificate. We ran grafana on container based on the official image pulled form the repo. Is there any workaround for this?

**Comment by valdemarpavesi** [comment]
I had same problem using official docker  grafana/grafana.

 there is no permission to install new certification.

here is the solution:

install cert to HOST.

case ubuntu:

test by curl
# curl https://mattermost.com/hooks/cqcdo196cidnix5hrf17g6kbdh

copy cert file to:

# cd /usr/share/ca-certificates/grafana/

mattermost.crt  

format must be CRT.

copy using chrome:

Steps
1. Open Google Chrome. ...
2. Select Show Advanced Settings > Manage Certificates.
3. Select the certificate you wanted to export then click "Export" button then next.
4. Now, you will get a "Certificate Export Wizard" box. ...
5. Select "No, do not export the private key" then click next.
6. Select "DER encoded binary x.509(.cer) then click next.


format is CER and must be converted to CRT

# openssl x509 -inform DER -in  mattermost.cer -out  mattermost.crt 


on mattermost.crt  check for

"Certificate Authority Information Access"

my case was :  DigiCertBaltimoreCA-2G2.crt

install certificates:

# dpkg-reconfigure ca-certificates

all alias installed to
# ls /etc/ssl/certs


found out that had to install DigiCertBaltimoreCA-2G2.crt,  by strace 

# strace  curl https://mattermost.com/hooks/cqcdo196cidnix5hrf17g6kbdh


openssl based on cert required will build a hash to find  installed cert

 stat("/etc/ssl/certs/8ad296c9.0", 0x7ffcdee57550) = -1 ENOENT (No such file or directory)

but,  from mattermost.crt   hash is:  cb318169

# openssl x509 -hash -fingerprint -noout -in /usr/share/ca-certificates/grafana/mattermost.crt
cb318169
SHA1 Fingerprint=04:25:3C:72:0C:95:9D:A2:9C:1E:4F:AA:55:C6:6F:94:B9:8F:25:07

+++

it was missing the root certification.


#  openssl x509 -hash -fingerprint -noout -in /usr/share/ca-certificates/grafana/DigiCertBaltimoreCA-2G2.crt
8ad296c9
SHA1 Fingerprint=A9:D5:30:02:E9:7E:00:E0:43:24:4F:3D:17:0D:6F:4C:41:41:04:FD

and ROOT certification have   hash 8ad296c9.


had two install both: mattermost.crt   + DigiCertBaltimoreCA-2G2.crt


++++

now, cert is working

# curl https://mattermost.com/hooks/cqcdo196cidnix5hrf17g6kbdh
<!DOCTYPE html> <html> <head> <meta http-equiv=X-UA-Compatible content="IE=edge"> <meta name=viewport content="width=device-width,initial-scale=1,maximum-scale=1,user-scalable=0"> <meta name=robots content="noindex, nofollow"> <meta name=referrer 

++++

now, have to inform docker to use HOST CERT from /etc/ssl/certs

mount host dir /etc/ssl/certs:

# docker run -d --name=grafanamm --hostname=grafanamm  -p 3000:3000  -v /etc/ssl/certs:/etc/ssl/certs -e "GF_INSTALL_PLUGINS=grafana-piechart-panel"   grafana/grafana

test inside the docker:

# docker exec -it grafana bash
bash-5.0$   curl https://mattermost.com/hooks/cqcdo196cidnix5hrf17g6kbdh

<!DOCTYPE html> <html> <head> <meta http-equiv=X-UA-Compatible content="IE=edge"> <meta name=viewport content="width=device-width,initial-scale=1,maximum-scale=1,user-scalable=0"> <meta name=robots content="noindex, nofollow"> <meta name=referrer 

++++
working.

**Comment by dmpe** [comment]
Facing same problem. Could you please at least add an option to disable TLS /SSL check - I am also using self-signed certs and this is very hard to deal with.

**Comment by bs444** [comment]
Bump.

**Comment by switchgl** [comment]
Bump, preferably an option within the notification channel options to tinker with cert details.

**Comment by bapBardas** [comment]
Same issue faced here using the docker image grafana/grafana when trying to configure a notification channel via webhook.

**Comment by bapBardas** [comment]
I similarly had the issue where my webhook notification channel was targetting a https exposed REST server using a certificate signed by an untrusted root certificate.

I managed to have this issue fixed by mounting a self written ca-certificates.crt file in the grafana container (which is based on debian FYI).

Based on those inputs :
- https://github.com/kubernetes-retired/heapster/issues/1653
- https://github.com/golang/go/blob/master/src/crypto/x509/root_linux.go#L9

I mounted a configmap file as a volume in /etc/ssl/certs/ in my kuberntes pod so that the ca-certificates.crt file contains the root certificate used to sign the https certificate.

**Comment by mdegat01** [comment]
I think I'm seeing this? It seems related but not sure if its a separate issue so figured I'd ask. My alert is non-functional because I see this every time it runs its query to see if it needs to alert me:

```
2021/04/01 11:03:31 https://39bd2704-loki:3100/loki/api/v1/query_range?direction=BACKWARD&end=1617289411880222719&interval=1&limit=1000&query=count_over_time%28%28%7Bcontainer_name%3D%22hassio_supervisor%22%7D+%7C~+%22Watchdog+found+a+problem+with%22%29%5B1m%5D%29&start=1617289351880222719&step=1
t=2021-04-01T11:03:31-0400 lvl=eror msg="Alert Rule Result Error" logger=alerting.evalContext ruleId=1 name="Addon crashed" error="tsdb.HandleRequest() error Get \"https://39bd2704-loki:3100/loki/api/v1/query_range?direction=BACKWARD&end=1617289411880222719&interval=1&limit=1000&query=count_over_time%28%28%7Bcontainer_name%3D%22hassio_supervisor%22%7D+%7C~+%22Watchdog+found+a+problem+with%22%29%5B1m%5D%29&start=1617289351880222719&step=1\": x509: certificate signed by unknown authority" changing state to=alerting
```

But this only happens when the query is run by the alert. If I go to explore and run this same query it runs without issue:

```
[01/Apr/2021:11:09:09 -0400] 200 192.168.1.11, 192.168.1.5, 172.30.32.1(172.30.32.2) GET /api/datasources/proxy/3/loki/api/v1/query_range?direction=BACKWARD&limit=1000&query=%7Bcontainer_name%3D%22hassio_supervisor%22%7D+%7C~+%22Watchdog+found+a+problem+with%22&start=1617286149000000000&end=1617289750000000000&step=1 HTTP/1.1 (Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:86.0) Gecko/20100101 Firefox/86.0)
```

It looks like the alert component is running queries differently? It's missing the `datasources/proxy/3` at the front of the URL it calls to do the query.

**Comment by large64** [comment]
@mdegat01 I've just managed to solve this problem for myself. It had nothing to do with my Grafana setup. My source of data is an influxdb Docker container which happened to use the wrong cert file. I'm using certbot to get my certs ready on my host. I've been using `cert.pem` file as `INFLUXDB_HTTP_HTTPS_CERTIFICATE` env variable in my container. Then I've read README (/etc/letsencrypt/live/README) which told me that
```
`[cert name]/privkey.pem`  : the private key for your certificate.
`[cert name]/fullchain.pem`: the certificate file used in most server software.
`[cert name]/chain.pem`    : used for OCSP stapling in Nginx >=1.3.7.
`[cert name]/cert.pem`     : will break many server configurations, and should not be used
                 without reading further documentation (see link below).
```
So I started using `fullchain.pem` instead of `cert.pem` and querying is working like a charm! I'm sooo happy! :)

**Comment by duchymor** [comment]
This worked for me on Alpine based Docker image:
`
USER root
COPY cacert.crt /usr/share/ca-certificates/custom/
WORKDIR /usr/share/ca-certificates/
RUN ls -1R custom/* >> /etc/ca-certificates.conf;
RUN update-ca-certificates
USER grafana
`

**Comment by relishcolouredhat** [comment]
Yeah, this being missing is a bit frustrating - we want to direct a webhook towards a service behind our k8s ingress; but get an error that the x509 cert is for the ingress and not the service hostname...

**Comment by mirkobrankovic** [comment]
It would be nice to have Skip TLS Verify option on Alert channels just like on Data sources :)

**Comment by gregbown** [comment]
Had this same issue running Grafana 9.4.7-1 on Rocky Linux release 8.7, trying to send an alert to a. NodeJS alert management tool. Both the Grafana app and the NodeJS app were using the self signed root via a cert trust chain. My development environment is set up to replicate production.

The only documentation I found that had marginal mention of this TLS verification were a few threads in community.
![grafana-ssl-error-in-rocky-linux](https://user-images.githubusercontent.com/1952728/232123380-3173e9ff-d284-4a2a-90c8-dd88b1b392df.jpg)

The solution was to update-ca-trust adding my root CA cert to ca-bundle in my Rocky Linux instance
```
sudo cp rootCAcert.pem /etc/pki/ca-trust/source/anchors/rootCAcert.pem
sudo update-ca-trust extract
```
**It would ease prototyping or development if the TLS verification was configurable**

Separately, I do not think there is any documentation on TLS verification

It seemed there are missing details on grafana.ini properties related to SSL and configuring the domain name
[The eluded to message](https://grafana.com/docs/grafana/latest/setup-grafana/set-up-https/#configure-grafana-https-and-restart-grafana) of running Grafana as a service with SSL on port 443 or <1024, configuration of  is only covered in [an obscure unrelated area](https://grafana.com/docs/grafana/latest/setup-grafana/installation/debian/)

**Comment by sbengo** [comment]
Hi, faced this on the new grafana unified alert feature.

The webhook contact point doesn't allow to configure the skip verify.
Are there any short plans to add this?
Should I create a new issue as it is related with the new unified alerting @armandgrillet ?

Thanks,
Regards

**Comment by BICH0** [comment]
> Hi, faced this on the new grafana unified alert feature.
> 
> The webhook contact point doesn't allow to configure the skip verify. Are there any short plans to add this? Should I create a new issue as it is related with the new unified alerting @armandgrillet ?
> 
> Thanks, Regards

I totally agree with this but, since this issue is 7 years old, it seems like it wont be happening soon :cry:

**Comment by jtackaberry** [comment]
Ran into this over here.  Absent building a custom container image, will use `skip_verify` as a crutch.

The most obvious way to implement this would be to add an `smtp.ca_cert_path` value, similar to what exists for the `database` section.

**Comment by tmalaher-telus** [comment]
I don't have root privs on this Linux box that is running my grafana. But I can modify anything else (environment, grafana.ini, invocation command line, ...).
Is there really no way of telling the notification webhooks to use a custom ca-certs file?
Why is there no global `ca_cert_path` and/or `tls_skip_verify_insecure` option.
Each sub-section (smtp, oauth) could have its own overrides if the global ones are not appropriate. 

I also don't have docker/podman installed on this box, otherwise I'd just use the solution that "mounts" the ca-certs file within a container to get around the lack of root privs.

**Comment by gzymomo** [comment]
This worked for me on k8s:

1.Store the certificate files into a Secret named **cert-files**
2.Create grafana deployment

...
containers:
  - args:
      - "/usr/sbin/update-ca-certificates && /run.sh"
    command:
      - /bin/sh
      - '-c'
    ...
    volumeMounts:
      - mountPath: /usr/local/share/ca-certificates/
        name: custom-cert
      - mountPath: /etc/ssl/certs
        name: installed-dir
initContainers:
  - command:
      - chmod
      - 'R'
      - '775'
      - /etc/ssl/certs
    image: ..../busybox:1.31.1
    ...
    volumeMounts:
      - mountPath: /etc/ssl/certs
        name: installed-dir
 ...
volumes:
  - name: custom-cert
    secret:
      defaultMode: 420
      secretName: cert-files
  - emptyDir:
      medium: Memory
      sizeLimit: 10Mi
    name: installed-dir
...

**Comment by titolins** [comment]
Re-opening this since we didn't make the 11.3.x release. The PR has been merged though so it's available in main and will be available on versions > 11.4
